cpp/src/binary/jit/util/type.cpp

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Copyright 2018-2019 BlazingDB, Inc.
+ *     Copyright 2018 Christian Noboa Mardini <christian@blazingdb.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "binary/jit/util/type.h"
+#include "utilities/type_dispatcher.hpp"
+
+namespace cudf {
+namespace binops {
+namespace jit {
+
+    /**---------------------------------------------------------------------------*
+     * @brief Functor to get type name in string
+     * 
+     * This functor uses the unofficial compiler macro __PRETTY_FUNCTION__
+     * to obtain the function signature which contains the template type name.
+     * The type name is searched and extracted from the string.
+     * 
+     * Example (clang): __PRETTY_FUNCTION__ =
+     * std::string type_name::operator()() [T = short]
+     * returns std::string("short")
+     * 
+     * Example (gcc): __PRETTY_FUNCTION__ =
+     * std::__cxx11::string type_name::operator()() [with T = short int; std::__cxx11::string = std::__cxx11::basic_string<char>]
+     * returns std::string("short int")
+     * 
+     * In case the type is wrapped using `wrapper`, the extra string "wrapper<" is
+     * also skipped to get the underlying wrapped type.
+     * 
+     *---------------------------------------------------------------------------**/
+    struct type_name {
+        template <class T>
+        CUDA_HOST_DEVICE_CALLABLE
+        std::string operator()() {
+#if defined(__clang__) || defined(__GNUC__)
+            std::string p = __PRETTY_FUNCTION__;
```

Comment:
This is clever, but something like this: https://stackoverflow.com/a/4485051/749748 might be clearer and less bug prone (also faster at run time).  Could even build this functionality into `cudf::type_dispatcher` so it's easily available if and when we add more JITs. @jrhemstad thoughts?

Reply:
@jrhemstad's [thoughts](https://github.com/rapidsai/cudf/pull/892/files/b6650ce3d0e327aee018a860078d546d8025f94b#r258271947) about this.

Reply:
My preference is to ruthlessly curtail the amount of code that needs to be manually updated when dtypes are added/changed, and that in an ideal world, the only thing that need to be changed are the body of the `type_dispatcher`.

In order to satisfy that desire, that did unfortunately require a bit of messy looking code. If @harrism feels strongly that manually maintained, but less messy code is preferred to what @devavret has now, then I wont argue against changing it.